File: rplugin/python3/deoplete/source/deoppet.py
# =============================================================================
# FILE: deoppet.py
# AUTHOR: Shougo Matsushita <Shougo.Matsu at gmail.com>
# License: MIT license
# =============================================================================

import re

from deoplete.base.source import Base


class Source(Base):
    def __init__(self, vim):
        Base.__init__(self, vim)

        self.name = 'deoppet'
        self.mark = '[dp]'
        self.rank = 200

    def gather_candidates(self, context):
        bvars = self.vim.current.buffer.vars
        if 'deoppet_snippets' not in bvars or not isinstance(
                bvars['deoppet_snippets'], dict):
            return []

        m1 = re.match(r'\w+$', context['complete_str'])
        m2 = re.match(r'\S+$', context['complete_str'])
        snippets = bvars['deoppet_snippets'].values()
        if m1 and m2 and m1.group(0) != m2.group(0):
            snippets = [x for x in snippets if x['options']['word']]
        ret = []
        for x in snippets:
            menu = x.get('abbr') or x.get('text')
            snippet_candidates = [
                {'word': w, 'menu': menu}
                for w in [x['trigger']] + x.get('alias', [])
            ]
            if x['options'].get('head'):
                m = re.search(r'^\s*(\S+)$', context['input'])
                if m and m.group(1) in x['trigger']:
                    ret += snippet_candidates
            elif not x['regexp'] or self.vim.call(
                    'matchstr', context['input'], x['regexp']):
                ret += snippet_candidates
        return ret

File: denops/ddc-sources/deoppet.ts
import {
  BaseSource,
  Candidate,
  Context,
} from "https://deno.land/x/ddc_vim@v0.5.0/types.ts";
import {
  Denops,
  fn,
  vars,
} from "https://deno.land/x/ddc_vim@v0.5.0/deps.ts#^";

export class Source extends BaseSource {
  async gatherCandidates(args: {
    denops: Denops,
    context: Context,
  }): Promise<Candidate[]> {
    const snippets = await vars.b.get(args.denops, "deoppet_snippets") as Record<
      string,
      unknown
    >[];
    if (!snippets) {
      return [];
    }

    const wordMatch = /\w+$/.exec(args.context.input);
    const charsMatch = /\S+$/.exec(args.context.input);
    const isWord = wordMatch && charsMatch && wordMatch[0] == charsMatch[0];

    const ret: Record<string, Candidate> = {} as Record<string, Candidate>;
    for (const key in snippets) {
      const val = snippets[key];
      const menu = val.abbr
        ? (val.abbr as string)
        : (val.text as string).replaceAll(/\n/g, "");

      const triggerCandidates = [val.trigger].concat(val.alias ? val.alias : [])
        .map((v) => ({
          word: v as string,
          menu: menu,
        }));

      const options = val.options as Record<string, Candidate>;
      if (
        (!options.head || /^\s*\S+$/.test(args.context.input)) &&
        (!options.word || isWord) &&
        (!val.regexp ||
          await fn.matchstr(args.denops, args.context.input, val.regexp) != "")
      ) {
        for (const candidate of triggerCandidates) {
          const word = candidate.word as string;
          if (!(word in ret)) {
            ret[word] = candidate;
          }
        }
      }
    }
    return Object.values(ret);
  }
}
